File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res, tmp = [], []
        candidates.sort()

        def backtrack(candidates, target, idx, curr_sum):
            if curr_sum == target:
                res.append(copy.deepcopy(tmp))
                return
            
            if curr_sum > target or idx == len(candidates):
                return

            for i in range(idx, len(candidates)):
                if(i>idx and candidates[i] == candidates[i-1]):
                    continue
                tmp.append(candidates[i])
                backtrack(candidates, target, i+1, curr_sum+candidates[i])
                tmp.pop()

        backtrack(candidates, target, 0, 0)
        
        return res 

File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> res = new ArrayList<>();
        backtrack(candidates, target, 0, 0, new ArrayList<>(), res);
        return res;
    }

    private void backtrack(int[] candidates, int target, int start, int sum, List<Integer> curr, List<List<Integer>> res) {
        if (sum == target) {
            res.add(new ArrayList<>(curr));
            return;
        }
        for (int i = start; i < candidates.length; i++) {
            if (i > start && candidates[i] == candidates[i - 1]) continue;
            if (sum + candidates[i] > target) break;
            curr.add(candidates[i]);
            backtrack(candidates, target, i + 1, sum + candidates[i], curr, res);
            curr.remove(curr.size() - 1);
        }
    }
}

